README.md
Planetarium
===========

Sub-pixel precision light spot rendering library for astronomy
and video tracking applications.

Sample image
------------

![Sample](https://raw.githubusercontent.com/ravenexp/planetarium/main/tests/test_8bpp.png)

Example usage
-------------

```rust
use planetarium::{Canvas, SpotShape};

// Draw on a square 256x256 pixel canvas.
let mut c = Canvas::new(256, 256);

// Define a round spot shape with diffraction radius of 2.5 pixels.
let shape = SpotShape::default().scale(2.5);

// Add some spots at random positions with varying shape size
// and peak intensity.
let spot1 = c.add_spot((100.3, 130.8), shape, 0.5);
let spot2 = c.add_spot((80.6, 200.2), shape.scale(0.5), 0.9);

// Note: Out of range position coordinates and peak intensities are fine.
//       The resulting spot image is clipped into the canvas rectangle.
//       Peak intensity > 1.0 leads to saturation to the maximum pixel value.
let spot3 = c.add_spot((256.1, 3.5), shape.scale(10.0), 1.1);

// Set the canvas background pixel value.
c.set_background(100);

// Clear the canvas and paint the light spots.
c.draw();

// Access the rendered image data as a linear pixel array.
let image_pixbuf = c.pixels();

// Get pixel at x = 100, y = 200.
let (x, y) = (100, 200);
let (image_width, image_height) = c.dimensions();
let val_x_y = image_pixbuf[(y * image_width + x) as usize];
```

Light spot parameters adjustment
--------------------------------

Some of the light spot parameters like coordinates and peak intensity
can be adjusted after the spot has been added to the canvas.

The spot position coordinates can be changed by adding an offset vector
and the peak intensity can be adjusted by multiplying with a spot
illumination factor.

It is possible to define a custom world coordinates to canvas coordinates
transformation, which affects all light spots.

```rust
use planetarium::{Canvas, SpotShape, Transform};

// Draw on a square 256x256 pixel canvas.
let mut c = Canvas::new(256, 256);

// Define an elliptic spot shape with diffraction radii of 2.5 x 1.5 pixels
// rotated by 45 degrees counter-clockwise.
let shape1 = SpotShape::default().stretch(2.5, 1.5).rotate(45.0);

// Define an elliptic spot shape by a 2x2 linear transform matrix.
let shape2 = SpotShape::from([[2.0, -0.5], [1.5, 3.0]]);

// Add some spots at random positions with varying shape size
// and peak intensity.
let spot1 = c.add_spot((100.3, 130.8), shape1, 0.5);
let spot2 = c.add_spot((80.6, 200.2), shape2, 0.9);

// Shift the rendered spot positions by applying the relative offset vectors.
// The intrinsic spot position coordinates are immutable.
c.set_spot_offset(spot1, (-34.2, 12.6));
c.set_spot_offset(spot2, (114.2, -73.3));

// Adjust the rendered spot peak intensity by applying the spot illumination factors.
// The intrinsic spot intensities are immutable.
c.set_spot_illumination(spot1, 1.2);
c.set_spot_illumination(spot2, 0.7);

// Query the resulting spot coordinates on the canvas.
let pos1 = c.spot_position(spot1).unwrap();
let pos2 = c.spot_position(spot2).unwrap();

// Query the resulting peak spot intensities.
let int1 = c.spot_intensity(spot1).unwrap();
let int2 = c.spot_intensity(spot2).unwrap();

// Apply a custom world coordinates to canvas coordinates transformation.
c.set_view_transform(Transform::default().translate((13.7, -20.3)));

// Query the resulting spot coordinates on the canvas after
// the view coordinate transformation.
let pos1x = c.spot_position(spot1).unwrap();
let pos2x = c.spot_position(spot2).unwrap();
```

Canvas image export
-------------------

The `Canvas` object supports image export to RAW and PNG file formats.
Both 8-bit and 16-bit PNG sample formats are supported.
Export to PNG formats requires the default `png` feature to be enabled.

### Example image export code

```rust
let mut c = Canvas::new(256, 256);

c.set_background(1000);
c.clear();

// Export to a 8-bit gamma-compressed grayscale RAW image.
let raw_8bpp_bytes = c.export_image(ImageFormat::RawGamma8Bpp).unwrap();

// Export to a 10-bit linear light grayscale little-endian RAW image.
let raw_10bpp_bytes = c.export_image(ImageFormat::RawLinear10BppLE).unwrap();

// Export to a 12-bit gamma-compressed grayscale little-endian RAW image.
let raw_12bpp_bytes = c.export_image(ImageFormat::RawLinear12BppLE).unwrap();

// Export to a 8-bit gamma-compressed grayscale PNG image.
let png_8bpp_bytes = c.export_image(ImageFormat::PngGamma8Bpp).unwrap();

// Export to a 16-bit linear light grayscale PNG image.
let png_16bpp_bytes = c.export_image(ImageFormat::PngLinear16Bpp).unwrap();
```

Window image export
-------------------

The `Canvas` object additionally supports windowed image export.

A single rectangular window represents a region of interest (ROI)
on the canvas image. Window rectangle coordinates are represented
by the public `Window` structure.

### Example window image export code

```rust
let c = Canvas::new(256, 256);

// Create a 32x16 pixels window with origin at (100, 150).
let wnd = Window::new(32, 16).at(100, 150);

let fmt = ImageFormat::RawGamma8Bpp;

// Export to the canvas window image bytes.
let raw_window_bytes = c.export_window_image(wnd, fmt).unwrap();
```

Subsampled image export
-----------------------

The `Canvas` object additionally supports subsampled image export
with independent row and column subsampling factors.

Only whole canvas images can be exported with subsampling.

### Example subsampled image export code

```rust
let c = Canvas::new(256, 256);

let fmt = ImageFormat::RawLinear10BppLE;

// Column (X) and row (Y) subsampling factors
let factors = (4, 2);

// Export to the subsampled canvas image bytes.
let raw_sub_bytes = c.export_subsampled_image(factors, fmt).unwrap();
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [0.2.0] - 2023-03-16

### Features

- [**breaking**] Use a larger pattern LUT (1024 entries)
- Add zero padding to the pattern LUT

### Miscellaneous Tasks

- Regenerate `cliff.toml` config

## [0.1.6] - 2023-03-13

### Documentation

- Add `Errors` section to function docstrings

### Features

- Perform subsampling rate factor validation

### Miscellaneous Tasks

- Silence remaining pedantic Clippy warnings

### Refactor

- Clean up the PNG image export code
- Add `must_use` function attributes
- Use inline string interpolation
- Fix pedantic Clippy nitpicks

## [0.1.5] - 2022-04-30

### CI configuration

- Update and simplify GitLab CI config
- Add `rust.yml` workflow
- Add `publish.yml` workflow

### Documentation

- Add a sample image to README.md
- Add subsampled image export examples

### Features

- Implement subsampled gamma RAW image export
- Implement subsampled 8-bit PNG image export
- Implement subsampled 16-bit PNG image export

### Testing

- Refactor canvas image export unit tests
- Add raw image export integration tests
- Add PNG image export integration tests

## [0.1.4] - 2022-03-03

### Bug Fixes

- Update the documentation publish job

### Features

- Add subsampled canvas image export method

## [0.1.3] - 2021-12-17

### CI configuration

- Fail the clippy check on warnings

### Documentation

- Add window image export examples

### Features

- Add a public rectangle struct named `Window`
- Add `window_spans()` method for windowed access
- Add the window image export method

### Refactor

- Move image export code into a submodule

## [0.1.2] - 2021-11-23

### CI configuration

- Add clippy checks with `--no-default-features`

### Documentation

- Add a `Canvas` struct doctest

### Features

- Improve `SpotShape` type ergonomics
- Implement the view coordinate transformation
- Improve `Transform` type ergonomics

## [0.1.1] - 2021-11-16

### CI configuration

- Add a basic GitLab CI config file

### Documentation

- Add spot parameters adjustment doctests
- Update README.md

### Features

- Add methods for spot parameters evaluation
- Implement export to RAW image formats

### Miscellaneous Tasks

- Add `git-cliff` config file
- Add a change log file

### Refactor

- Extract the pattern LUT into its own type
- Simplify the inner render loop

## [0.1.0] - 2021-11-08

### Bug Fixes

- Adjust effective spot radius estimation
- Rewrite spot image bounding box calculation

### Documentation

- Add a doctest for the main crate module
- Add PNG image export doctests
- Add some example code to README

### Features

- Add initial PNG image export support

### Miscellaneous Tasks

- Improve `SpotShape` type ergonomics

<!-- generated by git-cliff -->


src/export/raw.rs
//! Planetarium
//! ===========
//!
//! Private RAW image export routines
//! ---------------------------------
//!
//! Contains implementations of private methods
//! for the existing public types.

use crate::{Canvas, EncoderError, Window};

#[allow(clippy::unnecessary_wraps)]
impl Canvas {
    /// Exports the canvas window contents in the 8-bit gamma-compressed RAW image format.
    pub(super) fn export_raw8bpp(&self, window: Window) -> Result<Vec<u8>, EncoderError> {
        // Memory buffer to encode the RAW pixel data to
        let mut rawbuf: Vec<u8> = Vec::with_capacity(window.len());

        // The window is bounds checked by the caller.
        for span in self.window_spans(window).unwrap() {
            rawbuf.extend(span.iter().map(|p| self.gamma_curve.transform(*p)));
        }

        Ok(rawbuf)
    }

    /// Exports the canvas window contents in the `X`-bit linear light grayscale
    /// little-endian RAW image format.
    ///
    /// The const generic `X` must be in the range from 9 to 16.
    pub(super) fn export_raw1xbpp<const X: u16>(
        &self,
        window: Window,
    ) -> Result<Vec<u8>, EncoderError> {
        // Memory buffer to encode the RAW pixel data to
        let mut rawbuf: Vec<u8> = Vec::with_capacity(2 * window.len());

        // The window is bounds checked by the caller.
        for span in self.window_spans(window).unwrap() {
            for p in span {
                let bytes = (p >> (16 - X)).to_le_bytes();
                rawbuf.extend_from_slice(&bytes);
            }
        }

        Ok(rawbuf)
    }

    /// Exports the subsampled canvas contents in the 8-bit gamma-compressed
    /// RAW image format.
    pub(super) fn export_sub_raw8bpp(&self, factors: (u32, u32)) -> Result<Vec<u8>, EncoderError> {
        // Subsampled image size in pixels
        let pixlen = self.pixbuf.len() / (factors.0 * factors.1) as usize;

        // Memory buffer to encode the RAW pixel data to
        let mut rawbuf: Vec<u8> = Vec::with_capacity(pixlen);

        for i in 0..(self.height / factors.1) {
            let loffset = (i * factors.1 * self.width) as usize;

            for j in 0..(self.width / factors.0) {
                let offset = loffset + (j * factors.0) as usize;
                let xval = self.gamma_curve.transform(self.pixbuf[offset]);
                rawbuf.push(xval);
            }
        }

        Ok(rawbuf)
    }

    /// Exports the subsampled canvas contents in the `X`-bit linear light grayscale
    /// little-endian RAW image format.
    ///
    /// The const generic `X` must be in the range from 9 to 16.
    pub(super) fn export_sub_raw1xbpp<const X: u16>(
        &self,
        factors: (u32, u32),
    ) -> Result<Vec<u8>, EncoderError> {
        // Subsampled image size in pixels
        let pixlen = self.pixbuf.len() / (factors.0 * factors.1) as usize;

        // Memory buffer to encode the RAW pixel data to
        let mut rawbuf: Vec<u8> = Vec::with_capacity(2 * pixlen);

        for i in 0..(self.height / factors.1) {
            let loffset = (i * factors.1 * self.width) as usize;

            for j in 0..(self.width / factors.0) {
                let offset = loffset + (j * factors.0) as usize;
                let bytes = (self.pixbuf[offset] >> (16 - X)).to_le_bytes();
                rawbuf.extend_from_slice(&bytes);
            }
        }

        Ok(rawbuf)
    }
}

#[cfg(test)]
mod tests {
    use crate::{ImageFormat, SpotShape};

    use super::*;

    /// Creates a 256x256 canvas image for all tests.
    fn mkimage() -> Canvas {
        let mut c = Canvas::new(256, 256);
        c.set_background(1000);

        let shape = SpotShape::default().scale(4.5);
        let shape2 = shape.stretch(1.7, 0.7).rotate(45.0);

        c.add_spot((100.6, 150.2), shape, 0.9);
        c.add_spot((103.8, 146.5), shape2, 0.5);

        c.draw();
        c
    }

    #[test]
    fn export_raw8bpp() {
        let img = mkimage().export_image(ImageFormat::RawGamma8Bpp).unwrap();
        assert_eq!(img.len(), 256 * 256);
        assert_eq!(img[0], 33);
        assert_eq!(img[150 * 256 + 100], 238);
    }

    #[test]
    fn export_sub_raw8bpp() {
        let img = mkimage()
            .export_subsampled_image((2, 2), ImageFormat::RawGamma8Bpp)
            .unwrap();
        assert_eq!(img.len(), 256 * 256 / 2 / 2);
        assert_eq!(img[0], 33);
        assert_eq!(img[(150 * 128 + 100) / 2], 238);
    }

    #[test]
    fn export_window_raw8bpp() {
        let wnd = Window::new(32, 16).at(90, 140);

        let img = mkimage()
            .export_window_image(wnd, ImageFormat::RawGamma8Bpp)
            .unwrap();
        assert_eq!(img.len(), wnd.len());
        assert_eq!(img[300], 185);
    }

    #[test]
    fn export_raw10bpp() {
        let img = mkimage()
            .export_image(ImageFormat::RawLinear10BppLE)
            .unwrap();
        assert_eq!(img.len(), 256 * 256 * 2);
        assert_eq!(img[0], 0x0F);
        assert_eq!(img[1], 0x00);
        assert_eq!(img[2 * (150 * 256 + 100)], 106);
        assert_eq!(img[2 * (150 * 256 + 100) + 1], 3);
    }

    #[test]
    fn export_sub_raw10bpp() {
        let img = mkimage()
            .export_subsampled_image((2, 2), ImageFormat::RawLinear10BppLE)
            .unwrap();
        assert_eq!(img.len(), 256 * 256 * 2 / 2 / 2);
        assert_eq!(img[0], 0x0F);
        assert_eq!(img[1], 0x00);
        assert_eq!(img[2 * (150 * 128 + 100) / 2], 106);
        assert_eq!(img[2 * (150 * 128 + 100) / 2 + 1], 3);
    }

    #[test]
    fn export_window_raw10bpp() {
        let wnd = Window::new(32, 16).at(90, 140);

        let img = mkimage()
            .export_window_image(wnd, ImageFormat::RawLinear10BppLE)
            .unwrap();
        assert_eq!(img.len(), 2 * wnd.len());
        assert_eq!(img[300 * 2], 240);
        assert_eq!(img[300 * 2 + 1], 1);
    }

    #[test]
    fn export_raw12bpp() {
        let img = mkimage()
            .export_image(ImageFormat::RawLinear12BppLE)
            .unwrap();
        assert_eq!(img.len(), 256 * 256 * 2);
        assert_eq!(img[0], 0x3E);
        assert_eq!(img[1], 0x00);
        assert_eq!(img[2 * (150 * 256 + 100)], 168);
        assert_eq!(img[2 * (150 * 256 + 100) + 1], 13);
    }

    #[test]
    fn export_sub_raw12bpp() {
        let img = mkimage()
            .export_subsampled_image((4, 2), ImageFormat::RawLinear12BppLE)
            .unwrap();
        assert_eq!(img.len(), 256 * 256 * 2 / 4 / 2);
        assert_eq!(img[0], 0x3E);
        assert_eq!(img[1], 0x00);
        assert_eq!(img[2 * (150 / 2 * 64 + 100 / 4)], 168);
        assert_eq!(img[2 * (150 / 2 * 64 + 100 / 4) + 1], 13);
    }
}
